pkg/devicerepository/grpc.go

```diff
@@ -0,0 +1,203 @@
+func (dr *DeviceRepository) assetURL(brandID, path string) string {
+	if path == "" || dr.config.AssetsBaseURL == "" {
+		return path
+	}
+	return strings.TrimRight(dr.config.AssetsBaseURL, "/") + "/vendor/" + brandID + "/" + path
+}
+
+// ListBrands implements the ttnpb.DeviceRepositoryServer interface.
+func (dr *DeviceRepository) ListBrands(ctx context.Context, req *ttnpb.ListEndDeviceBrandsRequest) (*ttnpb.ListEndDeviceBrandsResponse, error) {
+	if err := rights.RequireApplication(ctx, req.ApplicationIDs, ttnpb.RIGHT_APPLICATION_DEVICES_READ); err != nil {
+		return nil, err
+	}
+	if req.Limit == 0 {
+		req.Limit = 1000
+	}
+	response, err := dr.store.GetBrands(store.GetBrandsRequest{
+		Limit:   req.Limit,
+		Page:    req.Page,
+		OrderBy: req.OrderBy,
+		Paths:   withDefaultBrandFields(req.FieldMask.Paths),
+		Search:  req.Search,
+	})
+	if err != nil {
+		return nil, err
+	}
+	for _, brand := range response.Brands {
+		brand.Logo = dr.assetURL(brand.BrandID, brand.Logo)
+	}
+	grpc.SetHeader(ctx, metadata.Pairs("x-total-count", strconv.FormatUint(uint64(response.Total), 10)))
+	return &ttnpb.ListEndDeviceBrandsResponse{
+		Brands: response.Brands,
+	}, nil
+}
+
+var (
+	errBrandNotFound = errors.DefineNotFound("brand_not_found", "brand `{brand_id}` not found")
+)
+
+// GetBrand implements the ttnpb.DeviceRepositoryServer interface.
+func (dr *DeviceRepository) GetBrand(ctx context.Context, req *ttnpb.GetEndDeviceBrandRequest) (*ttnpb.EndDeviceBrand, error) {
+	if err := rights.RequireApplication(ctx, req.ApplicationIDs, ttnpb.RIGHT_APPLICATION_DEVICES_READ); err != nil {
+		return nil, err
+	}
+	response, err := dr.store.GetBrands(store.GetBrandsRequest{
+		BrandID: req.BrandID,
+		Paths:   withDefaultBrandFields(req.FieldMask.Paths),
+		Limit:   1,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(response.Brands) == 0 {
+		return nil, errBrandNotFound.WithAttributes("brand_id", req.BrandID)
+	}
+	brand := response.Brands[0]
+	brand.Logo = dr.assetURL(brand.BrandID, brand.Logo)
+	return brand, nil
+}
+
+// ListModels implements the ttnpb.DeviceRepositoryServer interface.
+func (dr *DeviceRepository) ListModels(ctx context.Context, req *ttnpb.ListEndDeviceModelsRequest) (*ttnpb.ListEndDeviceModelsResponse, error) {
+	if err := rights.RequireApplication(ctx, req.ApplicationIDs, ttnpb.RIGHT_APPLICATION_DEVICES_READ); err != nil {
+		return nil, err
+	}
+	if req.Limit == 0 {
+		req.Limit = 1000
+	}
+	response, err := dr.store.GetModels(store.GetModelsRequest{
+		BrandID: req.BrandID,
+		Limit:   req.Limit,
+		Page:    req.Page,
+		Paths:   withDefaultModelFields(req.FieldMask.Paths),
+		Search:  req.Search,
+		OrderBy: req.OrderBy,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, model := range response.Models {
+		if photos := model.Photos; photos != nil {
+			photos.Main = dr.assetURL(model.BrandID, photos.Main)
+			for idx, photo := range photos.Other {
+				photos.Other[idx] = dr.assetURL(model.BrandID, photo)
+			}
+		}
+	}
```

Comment:
Please extract this to avoid repeating this. Also it's likely that other assets are added here in the future

Reply:
Extract as in a separate function right?

Reply:
This was already being done twice